Cache compilers.compile() in coredata

mensinda opened the pull request:
This patch caches the results of compilers.compile() in coredata instead of a class variable. This way, the cached values can be used during reconfiguration. This improves the reconfiguration time significantly (5 seconds on my system for gst-build).

Some changes to the internal API where necessary to print the `(cached)` string. In fact, this refactor makes up the vast majority of the changes. In case the user feedback (`(cached)`) is not necessary, these changes can be easily reverted.

(Re)configure times for gst-build with the current master:

```
../meson/meson.py build  15.99s user 4.40s system 106% cpu 19.216 total
../meson/meson.py --reconfigure build  13.14s user 3.22s system 106% cpu 15.426 total
```

And here are the times with this patch:

```
../meson/meson.py build  15.57s user 4.37s system 106% cpu 18.796 total
../meson/meson.py --reconfigure build  9.08s user 1.54s system 101% cpu 10.419 total
```

Akaricchi commented:
This is very nice. My project is much smaller than gst-build, yet this patchset brought down reconfigure times from ~3 seconds (with clang) to ~1s - a threefold reduction. I'll keep testing this and see if I find any problems.

nirbheek commented:
We definitely want the `(cached)` status, because otherwise people can get confused.

mensinda commented:
@nirbheek Sure. However, I am not 100% happy with the way I have implemented it now. Unfortunately, I couldn't come up with a better solution than handing the cached attribute all the way down to the interpreter.

Please let me know if anyone has a better idea on how to solve this.

dcbaker commented on mesonbuild/compilers/compilers.py:
```diff
@@ -893,42 +891,42 @@ def __repr__(self):
         return self.default_suffix
 
-    def get_define(self, dname, prefix, env, extra_args, dependencies):
+    def get_define(self, dname, prefix, env, extra_args, dependencies) -> (str, bool):
```
that should be `typing.Tuple[str, bool]`

mensinda replied:
Fixed with the last commit

dcbaker commented on mesonbuild/compilers/compilers.py:
```diff
@@ -1097,16 +1095,17 @@ def _get_compile_output(self, dirname, mode):
 
     @contextlib.contextmanager
-    def compile(self, code, extra_args=None, mode='link', want_output=False):
+    def compile(self, code, extra_args=None, mode='link', want_output=False, cdata: coredata.CoreData = None):
```
the type is invalid, you've assigned None but the type says it must be a `coredata.CoreData` instance. Either this needs to be a positional argument or `typing.Optional[coredata.CoreData]`

nirbheek commented:
I can't think of a better way to bubble this info up to the interpreter.

However, another way to implement this might be to use deterministic command-lines for all compiler checks, and rely on `ccache` to cache it for us. It would save us a lot of trouble w.r.t. keeping track of cache invalidation. When we're storing something for a single run that's not a problem, but when we're saving it for reconfigure, that can be a problem.

The issue there would be that `ccache` is not available for all toolchains or platforms. 

Of course, the existence of `--wipe` exist helps with the approach in this PR. We can always say that that's what people should use to force a full reconfigure. I am also biased in favour of this PR because it will help immensely on Windows where `ccache` is not easy to setup or use.

mensinda commented:
@dcbaker When the compiler version changes, all tests are rerun since compiler command and the compiler version are used in the key tuple (around line 1104). Of course, this only works if the compiler version itself is not cached.

@nirbheek I am not sure if relying on ccache would improve the performance since there is an overhead for spawning new processes. For cache invalidation, it might also be possible to get the dependency information from the compiler (similar to a dep file maybe)?

Also, the two Azure CI errors seem to be unrelated.

nirbheek commented:
> I am not sure if relying on ccache would improve the performance since there is an overhead for spawning new processes.

Yes, that's true, the main advantage would be simpler code because we can hand off all that work to `ccache`.

> For cache invalidation, it might also be possible to get the dependency information from the compiler (similar to a dep file maybe)?

That should be possible, yes. We can store the depinfo for each compiler check and then check the validity of the cache (once per cache entry per meson run) before returning it. Filesystem queries are always much faster than a subprocess invocation.

mensinda commented:
Rebased and the CI is green.

@dcbaker I think I should have addressed all your points with the last commit. Are there any other concerns?

@nirbheek Is the cache invalidation really required? Doing dep file parsing, timestamp tracking, etc. would significantly impact the complexity and (probably) also the performance of meson.
Also, from what I understand the dependencies cache is also never invalidated (even if let's say the dependency is removed).
If we implement some sort of cache invalidation, I would suggest implementing a more general API that can also be used for dependencies.

jpakkane commented:
Fine by me but I defer final acceptance for 0.50.0 to @nirbheek. This PR gets a freeze exception in case this can not be decided today.

textshell commented on mesonbuild/compilers/c.py:
```diff
@@ -763,7 +764,7 @@ def has_function(self, funcname, prefix, env, *, extra_args=None, dependencies=N
             if val is not None:
                 if isinstance(val, bool):
-                    return val
+                    return val, True
```
If it's from the cross file, it's not cached.

mensinda replied:
I guess this is true. Changing it to `False`

textshell commented on mesonbuild/compilers/compilers.py:
```diff
@@ -1155,8 +1154,16 @@ def compile(self, code, extra_args=None, mode='link', want_output=False):
+                              'pid', 'returncode', 'stdo', 'stde', 'text_mode']
+                    todel = [x for x in vars(p).keys() if x not in tokeep]
+                    for i in todel:
+                        delattr(p, i)
```
I wonder if the output of Popen_safe should not always be copied to a new object with just the properties that are needed. That way this functions return would not depend on if cdata is None.

mensinda replied:
When writing this, I found that some parts of the return value of `Popen_safe` save where accesed down the road. So changing this would requrie having a really clos looking at the entire compiers subfolder.

textshell replied:
I was thinking of always reducing the object to only contain tokeep. Or do you suspect that parts of the code that have caching disabled are using even more attributes?

mensinda replied:
I knew when originally writing this, but to give a definite answer I would have to look through the entire compiler code again.

I only know that somewhere `compile` is called without coredata so that nothing is cached and that some other value than `returncode` is accessed.

textshell commented on mesonbuild/compilers/c.py:
```diff
@@ -1016,7 +1018,7 @@ def find_library_real(self, libname, env, extra_dirs, code, libtype):
             args = ['-l' + libname]
             largs = self.linker_to_compiler_args(self.get_allow_undefined_link_args())
-            if self.links(code, env, extra_args=(args + largs)):
+            if self.links(code, env, extra_args=(args + largs))[0]:
```
I think this needs to by-pass the persistent cache, because find_library may only cache positive results.

mensinda replied:
Done

textshell commented:
Some last minute review fixes got integrated. Caching in meson is always a bit dangerous without further cache invalidation work (this holds for other caches too)

has_function with a prefix is likely used as a kind of version check in some cases. This will in the future need a `--wipe` as far as i understand. But the only way i can think of avoiding that is adding proper input tracking.

mensinda commented:
I have rebased this PR. @dcbaker I think I should have addressed all your points in your code review.

Are there any other open issues?

Ericson2314 commented:
I would say "what about cross compilers" but with https://github.com/mesonbuild/meson/pull/4010 it will be easy enough to generalize these things, so don't worry about it for now :).

mensinda commented:
Since this is the third rebase within a week, ping @jpakkane, @nirbheek, @dcbaker.

Ericson2314 commented on mesonbuild/coredata.py:
```diff
@@ -268,6 +268,7 @@ def __init__(self, options):
         self.compilers = OrderedDict()
         self.cross_compilers = OrderedDict()
         self.deps = OrderedDict()
+        self.compiler_check_cache = OrderedDict()
```
Can you annotate the keys to this with `# : ...`? Ideally we store everything that must be looked up from e.g. compiler options so we don't have to explicitly invalidate (bug prone) but just naturally get a cache miss if the key doesn't match.

mensinda replied:
The key is defined in `Compiler.compile` (`compilers.py:1138`). I am personally against duplicating such things across multiple files, especially because there won't be a unit test for this comment. Would you be OK with a comment pointing to the function in question?

Ericson2314 replied:
Oh silly me, I hadn't realized that we were already caching in the class, so there was already a caching strategy etc. Sorry!

jpakkane commented:
There is already a `--wipe` for `meson setup`. It should also wipe these settings if it does not already.

mensinda commented:
`--wipe` already works for this PR, since it removes the entire coredata file, where the compilation results are cached.

mensinda commented:
Rebased

jpakkane commented:
The coredata argument seems only used as `cdata.compiler_check_cache`. It would be clearer if the argument was `compiler_check_cache` directly instead. No point in having the additional `cdata` thing there.

mensinda commented:
@jpakkane Doing so would require to specify the cache explicitly in each invocation of the compile function (I am currently counting four that required it). This would seem like code duplication to me. This could lead to easy to miss bugs if we ever need to change something there again.

Would you be fine with moving the caching functionality out of the `compile` function into a `cached_compile` function? This way we can remove `cdata` from `compile` and the entire code should be more readable.

mensinda commented:
Rebased. I have also moved the caching functionality into a separate function. This way `cdata` isn't required in the `compile` function.

mensinda commented:
Thanks
